---
title: "Probiotics to the rescue: F. prausnitzii and beneficial communities to re-establish motor performance in Parkinson mouse models"
subtitle: "Data download and preparation"
author: "Anastasia Moiseyenko, Giacomo Antonello, Nicola Segata, Levi Waldron, Sarkis Mazmanian"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
params:
  basic_outdir: "results"
  microbiome_basic_transform: "RelAbund"
  tax_level: "SGB"
  beta_dist: "bray"
  beta_MDS: "MDS"
---

# Library preparation and `Google Cloud Storage` authentication

```{r, message=FALSE}
library(tidyverse) # install.packages("tidyverse")
library(TreeSummarizedExperiment) # BiocManager::install("TreeSummarizedExperiment")
library(biobakeryUtils) # remotes::install_github("g-antonello/biobakeryUtils")
library(parkinsonsMetagenomicData) # remotes::install_github("ASAP-MAC/parkinsonsMetagenomicData")

# googleCloudStorageR::gcs_auth(json_file = "~/Downloads/curatedmetagenomicdata-232f4a306d1d.json") # install.packages("googleCloudStorageR")

# try `parkinsonsMetagenomicData::output_file_types()` to find downloadable file types
```

# Load metadata

```{r}

metadata.df <- read_tsv("Data/metadata_5.0.tsv") %>% 
  # format a few values in the metadata
  mutate(
    genotype = as.factor(genotype) %>% relevel(ref = "WT"),
    cage = as.factor(cage),
    cohort = as.factor(cohort),
    subcohort = as.factor(subcohort),
    treatment = as.factor(treatment) %>% relevel(ref = "Control"), 
    treatment_legend = as.factor(gsub("F. prausnitzii", "*F. prausnitzii*", treatment)) %>% relevel(ref = "Control"), 
    genotype_treatment = paste(genotype, treatment) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO F. prausnitzii", "Thy1-ASO benCom")),
    genotype_treatment_legend = paste(genotype, treatment_legend) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO *F. prausnitzii*", "Thy1-ASO benCom"))
  )

rownames(metadata.df) <- metadata.df$uuid
```

# Prepare MetaPhlAn profiles as a `TreeSummarizedExperiment` object

```{r}

cache_table <- suppressMessages(cacheMetagenomicData(metadata.df$uuid, data_type = "relative_abundance", redownload = "no"))

MetaPhlAn.tse <- suppressMessages(loadMetagenomicData(cache_table))

# remove altExps, we don't need them
altExps(MetaPhlAn.tse) <- NULL

# replace the default metadata with the one that is important to us
## get uuid and number_reads
important_pieces_from_parkinsonsMetagenomicData <- MetaPhlAn.tse %>% 
  colData() %>% 
  as.data.frame() %>% 
  select(uuid, number_reads)

## generate a new colData data.frame with rows already ordered
new_colData <- as.data.frame(dplyr::left_join(metadata.df, important_pieces_from_parkinsonsMetagenomicData, by = "uuid"))
rownames(new_colData) <- new_colData$uuid
new_colData <- new_colData[colnames(MetaPhlAn.tse),]

## replace the old colData with the new one
colData(MetaPhlAn.tse) <- DataFrame(new_colData)

# create assay of Relative abundance as opposed to metaphlan's default percent
assay(MetaPhlAn.tse, "metaphlan_RelAbund") <- assay(MetaPhlAn.tse)/100

# order relative abundance [0,1]assay first 
assays(MetaPhlAn.tse) <- rev(assays(MetaPhlAn.tse))

# create a new assay with approximated counts as relAbund * reads_processed
assay(MetaPhlAn.tse, "metaphlan_Counts") <- t(round(apply(assay(MetaPhlAn.tse, "metaphlan_RelAbund"), 1, function(x) x * MetaPhlAn.tse@colData$number_reads), 0))

# add phylogenetic tree
MetaPhlAn.tse <-  AddPhyloTree_to_mpa_tse(data.tse = MetaPhlAn.tse, CHOCOPhlAn_version = "202403")

# rename strain column into SGB
colnames(rowData(MetaPhlAn.tse))[which(colnames(rowData(MetaPhlAn.tse)) == "strain")] <- "SGB" 

```

## Save R object, taxonomy and profiles tables in a separate folder

```{r}
write_TSE_to_dir(tse = MetaPhlAn.tse, out.dir = "Data/MetaPhlAn.tse")
# the following is not considered good practice, since the class is subject to 
# changes that may break the serialization, which is why I build the function 
# above
saveRDS(MetaPhlAn.tse, "Data/MetaPhlAn.tse/MetaPhlAn.tse.Rds")

tmp <- read_TSE_from_dir(tse.dir = "Data/MetaPhlAn.tse")
```

# Prepare HUMAnN gene families profiles as a `TreeSummarizedExperiment` object

```{r, message=FALSE}

cache_table_genefamilies_unstratified <- cacheMetagenomicData(metadata.df$uuid, data_type = "genefamilies_unstratified", redownload = "no") 
cache_table_genefamilies_relab_unstratified <- cacheMetagenomicData(metadata.df$uuid, data_type = "genefamilies_relab_unstratified", redownload = "no") 

HUMAnN_genefamilies_unstratified.tse <- suppressMessages(parkinsonsMetagenomicData::loadMetagenomicData(cache_table_genefamilies_unstratified))
HUMAnN_genefamilies_relab_unstratified.tse <- suppressMessages(parkinsonsMetagenomicData::loadMetagenomicData(cache_table_genefamilies_relab_unstratified))

# proof that we can re-derive other transformations done by humann
matrix1 <- assay(HUMAnN_genefamilies_unstratified.tse)
matrix2 <- assay(HUMAnN_genefamilies_relab_unstratified.tse)[rownames(matrix1), colnames(matrix1)]
identical(apply(matrix1, 2, function(x ) x/sum(x)), matrix2) # not identical...
range(apply(matrix1, 2, function(x) x/sum(x)) - matrix2) # but reeeeally close! 
# end of proof

## replace the default colData with the local one
assay(HUMAnN_genefamilies_unstratified.tse, "gene_family_abundance_rpk_relab_unstratified") <- apply(assay(HUMAnN_genefamilies_unstratified.tse), 2, function(x) x/sum(x))
assay(HUMAnN_genefamilies_unstratified.tse, "gene_family_abundance_rpk_cpm_unstratified") <- apply(assay(HUMAnN_genefamilies_unstratified.tse), 2, function(x) x/sum(x)*10^6)

colData(HUMAnN_genefamilies_unstratified.tse) <- DataFrame(metadata.df[colnames(HUMAnN_genefamilies_unstratified.tse),])

```

## Save R object and gene families profiles in a dedicated folder

```{r}
write_TSE_to_dir(HUMAnN_genefamilies_unstratified.tse, "Data")
# the following is not considered good practice, since the class is subject to 
# changes that may break the serialization, which is why I build the function 
# above
saveRDS(HUMAnN_genefamilies_unstratified.tse, "Data/HUMAnN_genefamilies_unstratified.tse/HUMAnN_genefamilies_unstratified.tse.Rds")
```

# Prepare HUMAnN pathway profiles as a `TreeSummarizedExperiment` object

```{r}
cache_table <- cacheMetagenomicData(metadata.df$uuid, data_type = "pathabundance_unstratified", redownload = "no") 

HUMAnN_pathways_unstratified.tse <- suppressMessages(parkinsonsMetagenomicData::loadMetagenomicData(cache_table))

# re-derive other transformations
assay(HUMAnN_pathways_unstratified.tse, "pathway_abundance_rpk_relab_unstratified") <- apply(assay(HUMAnN_pathways_unstratified.tse), 2, function(x) x/sum(x))
assay(HUMAnN_pathways_unstratified.tse, "pathway_abundance_rpk_cpm_unstratified") <- apply(assay(HUMAnN_pathways_unstratified.tse), 2, function(x) x/sum(x)*10^6)

## replace the default colData with the local one
colData(HUMAnN_pathways_unstratified.tse) <- DataFrame(metadata.df[colnames(HUMAnN_pathways_unstratified.tse),])

```

## Save R object and pathway profiles in a dedicated folder

```{r}
write_TSE_to_dir(HUMAnN_pathways_unstratified.tse, "Data")
# the following is not considered good practice, since the class is subject to 
# changes that may break the serialization, which is why I build the function 
# above
saveRDS(HUMAnN_pathways_unstratified.tse, "Data/HUMAnN_pathways_unstratified.tse/HUMAnN_pathways_unstratified.tse.Rds")
```

# SessionInfo

```{r}
sessionInfo()
```

