---
title: "Probiotics to the rescue: F. prausnitzii and beneficial communities to re-establish motor performance in Parkinson mouse models"
subtitle: "Data download and preparation"
author: "Anastasia Moiseyenko, Giacomo Antonello, Nicola Segata, Levi Waldron, Sarkis Mazmanian"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
params:
  metadata_version: "6.0"
---

# Library preparation and `Google Cloud Storage` authentication

```{r, message=FALSE}
library(tidyverse) # install.packages("tidyverse")
library(TreeSummarizedExperiment) # BiocManager::install("TreeSummarizedExperiment")
library(biobakeryUtils) # remotes::install_github("g-antonello/biobakeryUtils")
library(parkinsonsMetagenomicData) # remotes::install_github("ASAP-MAC/parkinsonsMetagenomicData")
library(tools)
# googleCloudStorageR::gcs_auth(json_file = "~/Downloads/curatedmetagenomicdata-232f4a306d1d.json") # install.packages("googleCloudStorageR")

# try `parkinsonsMetagenomicData::output_file_types()` to find downloadable file types
```

# Load metadata

```{r, eval=params$metadata_version=="5.0"}
metadata.df <- read_tsv("Data/metadata_5.0.tsv") %>% 
  # format a few values in the metadata
  mutate(
    genotype = as.factor(genotype) %>% relevel(ref = "WT"),
    cage = as.factor(cage),
    cohort = as.factor(cohort),
    subcohort = as.factor(subcohort),
    treatment = as.factor(treatment) %>% relevel(ref = "Control"), 
    treatment_legend = as.factor(gsub("F. prausnitzii", "*F. prausnitzii*", treatment)) %>% relevel(ref = "Control"), 
    genotype_treatment = paste(genotype, treatment) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO F. prausnitzii", "Thy1-ASO benCom")),
    genotype_treatment_legend = paste(genotype, treatment_legend) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO *F. prausnitzii*", "Thy1-ASO benCom"))
  ) %>% 
  as.data.frame()
```

```{r, eval=params$metadata_version=="6.0"}
metadata.df <- read_tsv("Data/metadata_6.0.tsv") %>% 
  # format a few values in the metadata
  mutate(
    genotype = as.factor(genotype) %>% relevel(ref = "WT"),
    cage = as.factor(cage),
    cohort = as.factor(cohort),
    subcohort = as.factor(subcohort),
    treatment = as.factor(treatment) %>% relevel(ref = "Control"), 
    treatment_legend = as.factor(gsub("F. prausnitzii", "*F. prausnitzii*", treatment)) %>% relevel(ref = "Control"), 
    genotype_treatment = paste(genotype, treatment) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO F. prausnitzii")),
    genotype_treatment_legend = paste(genotype, treatment_legend) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO *F. prausnitzii*"))
  ) %>% 
  as.data.frame()
```

```{r}
rownames(metadata.df) <- metadata.df$uuid
```

# Prepare MetaPhlAn profiles as a `TreeSummarizedExperiment` object

```{r}
avail_files <- output_file_types()
 
huggingFace_connection <- accessParquetData(data_types = c("relative_abundance", "pathabundance_unstratified"))

MetaPhlAn.tse <- loadParquetData(huggingFace_connection, data_type = "relative_abundance", filter_values = list(uuid = metadata.df$uuid))
# filter MetaPhlAn.tse to have only t__SGB level data, all columns of the Assay should sum to 100 at the 5th precision digit
# first add the exception of UNCLASSIFIED, which should be added as it is to the lowest levels
MetaPhlAn.tse <- MetaPhlAn.tse[!is.na(rowData(MetaPhlAn.tse)$clade_name_terminal) | rowData(MetaPhlAn.tse)$clade_name_kingdom == "UNCLASSIFIED",]
# then keep only the first useful columns and rename them, put the rest in metadata
new_rowData <- rowData(MetaPhlAn.tse) %>% 
  as.data.frame()

new_rowData_wanted_part <- select(new_rowData, clade_name_kingdom:clade_name_terminal) %>% 
  dplyr::rename_all(.funs = function(x) {
    gsub("clade_name_", "", x) %>% 
      tools::toTitleCase()
    }) %>% 
  dplyr::rename(SGB = Terminal)
# Fill UNCLASSIFIED row all with UNCLASSIFIED
new_rowData_wanted_part["UNCLASSIFIED",] <- paste(c("k", "p", "c", "o", "f", "g", "s", "t"), "UNCLASSIFIED", sep = "__")

# add new rowData to MetaPhlAn.tse and make shorter rownames
rowData(MetaPhlAn.tse) <- DataFrame(new_rowData_wanted_part)
rownames(MetaPhlAn.tse) <- rowData(MetaPhlAn.tse)[["SGB"]]

# create the extra_rowData dataframe to go into the metadata
extra_rowData <- select(new_rowData, !(clade_name_kingdom:clade_name_terminal))
rownames(extra_rowData) <- rownames(MetaPhlAn.tse)
metadata(MetaPhlAn.tse)$rowData_extraColumns <- extra_rowData

# put custom colData separately, but preserve the first 4 columns and use the first 1 to merge
rownames(metadata.df) <- metadata.df$uuid
old_colData <- as.data.frame(colData(MetaPhlAn.tse))

new_colData <- right_join(metadata.df, old_colData[,1:4], by = "uuid") %>% 
  mutate(
    number_reads = parse_number(reads_processed),
    reads_processed = NULL
  )
rownames(new_colData) <- new_colData$uuid

# replace the colData 
colData(MetaPhlAn.tse) <- DataFrame(new_colData)


# remodel the assays
assayNames(MetaPhlAn.tse) <- "percent"
assay(MetaPhlAn.tse, "relabundance") <- assay(MetaPhlAn.tse)/100
assay(MetaPhlAn.tse, "counts") <- t(apply(assay(MetaPhlAn.tse, "relabundance"), 1, function(x) x * MetaPhlAn.tse$number_reads))

# reorder the assays so that relative abundance (ranging from 0 to 1, not 0 to 100)
# is first.
assays(MetaPhlAn.tse) <- assays(MetaPhlAn.tse)[c(2,1,3)]

# add phylogenetic tree
MetaPhlAn.tse <-  AddPhyloTree_to_mpa_tse(data.tse = MetaPhlAn.tse, CHOCOPhlAn_version = "202403")
```

## Save R object, taxonomy and profiles tables in a separate folder

```{r}
write_TSE_to_dir(tse = MetaPhlAn.tse, out.dir = "Data/MetaPhlAn.tse")
# the following is not considered good practice, since the class is subject to 
# changes that may break the serialization, which is why I build the function 
# above
saveRDS(MetaPhlAn.tse, "Data/MetaPhlAn.tse/MetaPhlAn.tse.Rds")
```

# Prepare HUMAnN pathway profiles as a `TreeSummarizedExperiment` object

```{r}
HUMAnN.tse <- loadParquetData(huggingFace_connection, data_type = "pathabundance_unstratified", filter_values = list(uuid = metadata.df$uuid))

# fix rowData and rownames a little

new_rowData <- rowData(HUMAnN.tse) %>% 
  as.data.frame() %>%
  separate_wider_delim(cols = "pathway", delim = ": ", names = c("MetaCyc_code", "Description"), cols_remove = FALSE, too_few = "align_start") %>% 
  mutate(MetaCyc_code_safeName = make.names(MetaCyc_code)) %>% 
  relocate(pathway)
rowData(HUMAnN.tse) <- DataFrame(new_rowData)
# rename rownames for better compatibility
rownames(HUMAnN.tse) <- rowData(HUMAnN.tse)$MetaCyc_code_safeName

# put custom colData separately, but preserve the first 4 columns and use the first 1 to merge
rownames(metadata.df) <- metadata.df$uuid
old_colData <- as.data.frame(colData(HUMAnN.tse))

new_colData <- right_join(metadata.df, old_colData[,1:4], by = "uuid")
rownames(new_colData) <- new_colData$uuid

# replace the colData 
colData(HUMAnN.tse) <- DataFrame(new_colData)


# re-derive tranformations
assay(HUMAnN.tse, "relabundance") <- apply(assay(HUMAnN.tse), 2, function(x) x/sum(x))
assay(HUMAnN.tse, "cpm") <- apply(assay(HUMAnN.tse), 2, function(x) x/sum(x)*10^6)
```

## Save R object and pathway profiles in a dedicated folder

```{r}
write_TSE_to_dir(HUMAnN.tse, "Data/HUMAnN_pathways_unstratified.tse")
# the following is not considered good practice, since the class is subject to 
# changes that may break the serialization, which is why I build the function 
# above
saveRDS(HUMAnN.tse, "Data/HUMAnN_pathways_unstratified.tse/HUMAnN_pathways_unstratified.tse.Rds")
```

# Download and save Genome Taxonomy Data Base table from Metaphlan on Zenodo

```{r}
# load translation from SGB to GTDB taxonomy for further naming of bacteria
# this is taken from a Zenodo entry for metaphlan 4.2.2

download.file("https://zenodo.org/api/records/15594111/files-archive", destfile = "Metaphlan-4.2.2.zip", quiet = TRUE)
unzip("Metaphlan-4.2.2.zip", exdir = "Metaphlan-4.2.2")
unzip("Metaphlan-4.2.2/biobakery/MetaPhlAn-4.2.2.zip", exdir = "Metaphlan-4.2.2/")
GTDB.df <- read.delim("Metaphlan-4.2.2/biobakery-MetaPhlAn-500ce07/metaphlan/utils/mpa_vJun23_CHOCOPhlAnSGB_202403_SGB2GTDB.tsv", header = FALSE, se = "\t", col.names = c("SGB", "full_name")) %>% 
  relocate(full_name) %>% 
  separate_wider_delim(delim = ";", cols = "full_name", names = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), too_few = "align_start") %>% 
  mutate_all(function(x) gsub("d__|p__|c__|o__|f__|g__|s__", "", x)) 

write_tsv(GTDB.df, "Data/GTDB_202403.tsv")
unlink("Metaphlan-4.2.2*", recursive = TRUE)
```

# SessionInfo

```{r}
sessionInfo()
```

