---
title: "Probiotics to the rescue: F. prausnitzii and beneficial communities to re-establish motor performance in Parkinson mouse models"
subtitle: "Metaphlan 4.1.1 relative abundances"
author: "Anastasia Moiseyenko, Giacomo Antonello, Nicola Segata, Levi Waldron, Sarkis Mazmanian"
date: "`r Sys.Date()`"
output:
  rmdformats::readthedown:
    self_contained: true
    toc_depth: 3
    code_folding: hide
    toc_float: true
    number_sections: true
    thumbnails: false
    lightbox: true
    gallery: false
    use_bookdown: true
    highlight: haddock
params:
  basic_outdir: "results"
  assayTested: "metaphlan_RelAbund"
  microbiome_data_type: "SGB"
  beta_dist: "bray"
  beta_MDS: "MDS"
---

```{r}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

```{r, include=FALSE}
# packages for tabular data manipulation
library(tidyverse)
library(data.table)

# microbiome specific packages
library(maaslin3) # this version needs a custom fix in a fork I made pending approval as of Jul 28 2025. install with remotes::install_github("g-antonello/maaslin3", force = TRUE)
library(mia)
# some packages for data visualization (plots and tables)
library(miaViz)
library(pheatmap)
library(UpSetR)
library(ggpubr)
library(kableExtra)
library(reactable)
library(ggtext) # enhanced text handling in ggplot
# MAC's package to download data
library(parkinsonsMetagenomicData)

library(ggh4x) # for complex ggplot2 faceting
# remotes::install_github("cmartin/ggConvexHull")
library(ggConvexHull) # for convex hulls geometries
# set ggplot2 plot themes 
theme_set(theme_light())
# set seed for reproducibility
set.seed(1234)

library(biobakeryUtils)

joe_palette <- c("WT Control" = "#b2df8a", "Thy1-ASO Control" = "#a6cee3", "Thy1-ASO *F. prausnitzii*" = "#1f78b4", "Thy1-ASO F. prausnitzii" = "#1f78b4", "Thy1-ASO benCom" = "turquoise4")

```

# Create output directories all in one chunk

```{r}

output_directories.list <- list(
  # basic output directory for SGB
  baseDir = file.path(params$basic_outdir, params$microbiome_data_type),
  # alpha diversity directory
  alpha_div_outdir = file.path(params$basic_outdir, params$microbiome_data_type, "01_alpha_diversity"),
  # beta diversity output directory
  beta_div_outdir = file.path(params$basic_outdir, params$microbiome_data_type, "02_beta_diversity"),
  # cohort Fp7 directory
  cohort_Fp7_Dir = file.path(params$basic_outdir, params$microbiome_data_type, "cohort_Fp7")
)

tmp_out <- capture.output(sapply(output_directories.list, dir.create, showWarnings = FALSE, recursive = TRUE))
rm(tmp_out)
```

# Prepare data for analysis

This is run only if interactive to avoid careless overwriting

```{r, eval=interactive()}

# Load metadata
metadata.df <- read_tsv("Data/metadata_5.0.tsv") %>% 
  # format a few values in the metadata
  mutate(
    genotype = as.factor(genotype) %>% relevel(ref = "WT"),
    cage = as.factor(cage),
    cohort = as.factor(cohort),
    subcohort = as.factor(subcohort),
    treatment = as.factor(treatment) %>% relevel(ref = "Control"), 
    treatment_legend = as.factor(gsub("F. prausnitzii", "*F. prausnitzii*", treatment)) %>% relevel(ref = "Control"), 
    genotype_treatment = paste(genotype, treatment) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO F. prausnitzii", "Thy1-ASO benCom")),
    genotype_treatment_legend = paste(genotype, treatment_legend) %>% factor(levels = c("WT Control", "Thy1-ASO Control", "Thy1-ASO *F. prausnitzii*", "Thy1-ASO benCom"))
    )

googleCloudStorageR::gcs_auth(json_file = "~/Downloads/curatedmetagenomicdata-232f4a306d1d.json")

# METAPHLAN4
cache_table <- cacheMetagenomicData(metadata.df$uuid, data_type = "relative_abundance", redownload = "no") 

# metaphlan <- get_metaphlan_locators(metadata.df$uuid, data_type = "relative_abundance")

MetaPhlAn.tse <- suppressMessages(biobakeryUtils::loadMetagenomicData(cache_table))

# bypass the metadata including only what is important for us:

## get uuid and number_reads
important_pieces_from_parkinsonsMetagenomicData <- MetaPhlAn.tse %>% 
  colData() %>% 
  as.data.frame() %>% 
  dplyr::select(uuid, number_reads)

## generate a new colData data.frame with rows already ordered
new_colData <- as.data.frame(dplyr::left_join(metadata.df, important_pieces_from_parkinsonsMetagenomicData, by = "uuid"))
rownames(new_colData) <- new_colData$uuid
new_colData <- new_colData[colnames(MetaPhlAn.tse),]
## replace the old colData with the new one
colData(MetaPhlAn.tse) <- DataFrame(new_colData)

# create assay of Relative abundance as opposed to metaphlan's default percent
assay(MetaPhlAn.tse, "metaphlan_RelAbund") <- assay(MetaPhlAn.tse)/100
# order relative abundance [0,1]assay first 
assays(MetaPhlAn.tse) <- rev(assays(MetaPhlAn.tse))
# create a new assay with approximated counts as relAbund * reads_processed
assay(MetaPhlAn.tse, "metaphlan_Counts") <- t(round(apply(assay(MetaPhlAn.tse, "metaphlan_RelAbund"), 1, function(x) x * MetaPhlAn.tse@colData$number_reads), 0))

# add phylogenetic tree
MetaPhlAn.tse <-  AddPhyloTree_to_mpa_tse(data.tse = MetaPhlAn.tse, CHOCOPhlAn_version = "202403")

# rename strain column into SGB
colnames(rowData(MetaPhlAn.tse))[which(colnames(rowData(MetaPhlAn.tse)) == "strain")] <- "SGB" 

# calculate Firmicutes/Bacteroidetes Ratio
MetaPhlAn.tse@colData$BF.Ratio <- apply(assay(agglomerateByRank(MetaPhlAn.tse, "phylum")), 2, function(x){
  bactName <- "p__Bacteroidota"
  firmName <- "p__Firmicutes"
  return(x[bactName]/x[firmName])
})

MetaPhlAn.tse@colData$FB.Ratio <- apply(assay(agglomerateByRank(MetaPhlAn.tse, "phylum")), 2, function(x){
  bactName <- "p__Bacteroidota"
  firmName <- "p__Firmicutes"
  return(x[firmName]/x[bactName])
})

saveRDS(MetaPhlAn.tse, "Data/MetaPhlAn.tse.Rds")
```

```{r}
MetaPhlAn.tse <- readRDS("Data/MetaPhlAn.tse.Rds")
```

```{r, eval=interactive()}
# load translation from SGB to GTDB taxonomy for further naming of bacteria
# this is taken from a Zenodo entry for metaphlan 4.2.2

download.file("https://zenodo.org/api/records/15594111/files-archive", destfile = "~/Downloads/Metaphlan-4.2.2.zip", quiet = TRUE)
unzip("~/Downloads/Metaphlan-4.2.2.zip", exdir = "~/Downloads/Metaphlan-4.2.2")
unzip("~/Downloads/Metaphlan-4.2.2/biobakery/MetaPhlAn-4.2.2.zip", exdir = "~/Downloads/Metaphlan-4.2.2/")
GTDB.df <- read.delim("~/Downloads/Metaphlan-4.2.2/biobakery-MetaPhlAn-500ce07/metaphlan/utils/mpa_vJun23_CHOCOPhlAnSGB_202403_SGB2GTDB.tsv", header = FALSE, se = "\t", col.names = c("SGB", "full_name")) %>% 
  relocate(full_name) %>% 
  separate_wider_delim(delim = ";", cols = "full_name", names = c("Domain", "Phylum", "Class", "Order", "Family", "Genus", "Species"), too_few = "align_start") %>% 
  mutate_all(function(x) gsub("d__|p__|c__|o__|f__|g__|s__", "", x)) 

write_tsv(GTDB.df, "Data/GTDB_202403.tsv")
```

```{r}
GTDB.df <- read.delim("Data/GTDB_202403.tsv", colClasses = "character", sep = "\t" )
```

# Dataset-wide microbiome composition statistics

## Number of taxa per taxonomic level

```{r}
InputData_splitByRanks.tse <- append(splitByRanks(MetaPhlAn.tse), list("SGB" = MetaPhlAn.tse))
 
InputData_splitByRanks.tse %>%  
  sapply(nrow)
```

## Prevalence distribution of taxa per taxonomic rank

```{r}
InputData_splitByRanks.tse %>% 
  lapply(assay) %>% 
  lapply(apply, 1, function(x) sum(x > 0)) %>% 
  sapply(summary) %>% 
  kbl() %>% 
  kable_styling()
```

```{r}
n_taxa <- 20
```

## Relative abundance of `r n_taxa` most abundant bacteria per taxonomic level {.tabset}

```{r}

tmp <- InputData_splitByRanks.tse %>% 
  lapply(function(tse){
    max_n <- nrow(tse)
    topNMostAbund <- getTop(tse, assay.type = "metaphlan_RelAbund", method = "mean", top = min(c(n_taxa, max_n)))
    tse.subset <- tse[topNMostAbund,]
    return(tse.subset)
  })
  
```

```{r, results = "asis", fig.height=7, fig.width=10}

for (tx in names(tmp)) {
  # Create header with cat()
  cat("\n\n###", tx, "\n")
  
  # Create plot and convert to plotly
  p <- miaViz::plotAbundance(tmp[[tx]], 
                            assay.type = "metaphlan_RelAbund",
                            group = tx) +
       facet_nested(cols = vars(cohort, treatment), scales = "free_x") + 
    theme(legend.position = "bottom") + 
    guides(fill = guide_legend(nrow = 5))
  
  # Explicitly print plotly object
  print(p)
  
  cat("\n")
}

```

This is again to show that there are too many taxa to easily visualize them.
We need to rely on other methods.

## Heatmap all taxa and samples

```{r, fig.cap="Heatmap of full microbiome composition of the dataset. Relative abundances have been Z-scaled by taxon, to account for intrinsic differences in relative abundances in the microbiome"}

pheatmap::pheatmap(assay(MetaPhlAn.tse) %>% magrittr::set_colnames(colData(MetaPhlAn.tse)$sample_name), scale = "row", annotation_col = colData(MetaPhlAn.tse) %>% as.data.frame() %>%  magrittr::set_rownames(.$sample_name) %>% dplyr::select(treatment, cohort), angle_col = 315, show_rownames = FALSE)

```

# Analyses are performed at the `r params$microbiome_data_type` level

Below the `TreeSummarizedExperiment` object summary that is going to be used
in all following analyses

```{r}
MetaPhlAn.tse <- InputData_splitByRanks.tse[[params$microbiome_data_type]]

MetaPhlAn.tse
```

# Alpha diversity does not vary in relation to sequencing depth

```{r}
MetaPhlAn.tse <- addAlpha(MetaPhlAn.tse, assay.type = "metaphlan_RelAbund", index = c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"))

lapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), 
       function(div) {
         stats_basic <- shapiro.test(as.data.frame(colData(MetaPhlAn.tse))[[div]])
         text <- paste0("Shapiro-Wilk test: W = ", round(stats_basic$statistic, 2), "; P-value = ", round(stats_basic$p.value, 3))
         plot <- ggplot(as.data.frame(colData(MetaPhlAn.tse)), aes(x = !!sym(div))) +
           geom_histogram() + 
           labs(
             title = div, 
             caption = text
           )
         
         }
       ) %>% 
  ggarrange(plotlist = .)

```

```{r, tab.cap="Assessment of dependence of alpha diversity measures in relation to sequencing depth"}
tmp <- sapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), function(div) {
  lm(as.formula(paste(div, "~", "number_reads")), data = colData(MetaPhlAn.tse))
  
}, USE.NAMES = TRUE, simplify = FALSE)

table_results.df <- tmp %>% lapply(broom::tidy) %>% 
  bind_rows(.id = "AlphaIndex") %>% 
  filter(!grepl("Intercept", term))

table_results.df %>%
  kbl() %>% 
  row_spec(which(table_results.df$p.value < 0.05), bold = TRUE) %>% 
  kable_styling()
```

# Alpha diversity does not vary in relation to genotype, treatment and cohort

## Overall plot

```{r}
alpha_div_plotList <- lapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), function(div) {
  colData(MetaPhlAn.tse) %>% 
    as.data.frame() %>% 
  ggplot(aes(x = "", y = eval(parse(text = div)), color = genotype_treatment)) + 
  geom_point(position = position_jitterdodge(dodge.width = 0.8, seed = 1234)) + 
  geom_boxplot(alpha = 0.5, show.legend = FALSE, outliers = FALSE) + 
    facet_grid(cols = vars(cohort), scales = "free_y") +
    labs(
      title = div,
      x = NULL,
      y = "Alpha Estimate",
      color = "Genotype and Treatment"
    )})
```


```{r}
ggarranged_plotlist <-  ggarrange(plotlist = alpha_div_plotList, common.legend = TRUE)

ggsave(plot = ggarranged_plotlist, filename = file.path(output_directories.list$alpha_div_outdir, "alphas_vs_genotype_cohort_treatment.svg"), dpi = 600, height = 5, width = 6)
ggarranged_plotlist
```

## Statistics tables

### Both cohorts, adjusting for it

$$
AlphaIndex \sim genotype + treatment + cohort + \text{number_reads}
$$

```{r}
tmp <- sapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), function(div) {
  lm(as.formula(paste(div, "~", "genotype + treatment + cohort + number_reads")), data = colData(MetaPhlAn.tse))
  
}, USE.NAMES = TRUE, simplify = FALSE)

table_results.df <- tmp %>% lapply(broom::tidy) %>% 
  bind_rows(.id = "AlphaIndex") %>% 
  
  arrange(term) %>% 
  relocate(term)

write_tsv(table_results.df, file = file.path(output_directories.list$alpha_div_outdir, "alphas_vs_genotype_cohort_treatment.tsv"))

table_results.df %>%
  kbl() %>% 
  row_spec(which(table_results.df$p.value < 0.05), bold = TRUE, color = "red3") %>% 
  kable_styling()
```

### Only cohort `Fp7`

$$
AlphaIndex \sim genotype + treatment + \text{number_reads}
$$

```{r}
cohort_subdir <- "Fp7"

tmp <- sapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), function(div) {
  lm(as.formula(paste(div, "~", "genotype + treatment + number_reads")), data = colData(MetaPhlAn.tse)[colData(MetaPhlAn.tse)$cohort == cohort_subdir,])
  
}, USE.NAMES = TRUE, simplify = FALSE)

table_results.df <- tmp %>% lapply(broom::tidy) %>% 
  bind_rows(.id = "AlphaIndex") %>% 
  arrange(term) %>% 
  relocate(term)

write_tsv(table_results.df, file = file.path(params$basic_outdir, params$microbiome_data_type, paste("cohort",cohort_subdir, sep = "_"),  "alphas_vs_genotype_treatment.tsv"))

table_results.df %>%
  kbl() %>% 
  row_spec(which(table_results.df$p.value < 0.05), bold = TRUE, color = "red3") %>% 
  kable_styling()
```


### Only cohort `bCL`

$$
AlphaIndex \sim genotype + treatment + \text{number_reads}
$$

```{r}
cohort_subdir <- "bCL"

tmp <- sapply(c("dbp_dominance", "observed_richness", "shannon_diversity", "gini"), function(div) {
  lm(as.formula(paste(div, "~", "genotype + treatment + number_reads")), data = colData(MetaPhlAn.tse)[colData(MetaPhlAn.tse)$cohort == cohort_subdir,])
  
}, USE.NAMES = TRUE, simplify = FALSE)

table_results.df <- tmp %>% lapply(broom::tidy) %>% 
  bind_rows(.id = "AlphaIndex") %>% 
  arrange(term) %>% 
  relocate(term)

write_tsv(table_results.df, file = file.path(params$basic_outdir, params$microbiome_data_type, paste("cohort",cohort_subdir, sep = "_"),  "alphas_vs_genotype_treatment.tsv"))

table_results.df %>%
  kbl() %>% 
  row_spec(which(table_results.df$p.value < 0.05), bold = TRUE, color = "red3") %>% 
  kable_styling()
```


# Variables and potential confounders

Variables:

  * treatment: the probiotic community given, or the empty liquid Control gavaged
  * genotype: mice genotype, either WT or Thy1-ASO

Possible confounders:
  
  * Cohort: important to consider because it could determine inter-testing variability
  * Cage: Potential confounder, but mice receiving the same treatment are also in the same cage and there are only 2 mice per cage, making statics weak
  * Subcohort: Same as cage. It looks like a few samples were processed at a different time: 7W-F, 8W-F, 8W-B, 8C-B

# Beta diversity in relation to treatment and genotype

```{r}
dimRedName <- paste(params$beta_dist, "mds", sep = "_")
# this will work once mia or TreeSummarizedExperiment fixes some things
MetaPhlAn.tse <- mia::addMDS(
  MetaPhlAn.tse,
  assay.type = "metaphlan_RelAbund",
  method = params$beta_dist,
  ncomponents = 3, name = dimRedName
  )

variances_explained <- attr(reducedDim(MetaPhlAn.tse, dimRedName), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

project1_meta_with_MDS.df <- cbind.data.frame(as.data.frame(colData(MetaPhlAn.tse)), reducedDim(MetaPhlAn.tse, dimRedName) %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

beta_plot_basic <- project1_meta_with_MDS.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = genotype_treatment_legend)) + 
  geom_point(size = 2) + 
  stat_ellipse(show.legend = FALSE) + 
  scale_color_manual(values = joe_palette) +
  theme(legend.position = "top") +
  labs(color = "Genotype and Treatment",
       x = paste0("MDS 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("MDS 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

```

```{r}
ggsave(plot = beta_plot_basic, filename = file.path(output_directories.list$beta_div_outdir, "bray_MDS_plot.svg"), dpi = 600, height = 6.5, width = 7)

beta_plot_basic
```

## PERMANOVA statistics

```{r}
set.seed(1234)
MetaPhlAn.tse <- addPERMANOVA(
  MetaPhlAn.tse,
  assay.type = "metaphlan_RelAbund",
  method = "bray",
  formula = x ~ genotype + treatment + cohort,
  name = "permanova",
  permutations = 9999,
  test.homogeneity = TRUE
)

permanova_table.df <- PERMANOVA_to_table(MetaPhlAn.tse)

write_tsv(permanova_table.df, file = file.path(output_directories.list$beta_div_outdir, paste(params$beta_dist, "adonisPermanova", "genotype", "treatment", "cohort.tsv", sep = "_")))
```

```{r}
permanova_table.df %>% 
  kbl() %>% 
  kable_styling()
```

# Microbiome differences between genotypes in treatment control mice

```
Are Thy1-ASO mice having different microbiomes than WT, regardless of the experimental
cohort?
```

For this, I will include both cohorts together and adjust for cohort in the model.
Also, I will compare Control controls using WT as baseline.

$$
\log_2(Microbiome) ~ genotype + cohort
$$

```{r}
# prepare input data

tmp.tse <- MetaPhlAn.tse[,MetaPhlAn.tse@colData$treatment == "Control"] 
tmp.tse@colData$genotype <- relevel(tmp.tse@colData$genotype, "WT") # not needed, but better safe than sorry
tmp.tse@colData$genotype <- droplevels(tmp.tse@colData$genotype)

tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "genotype_vs_WT_ControlCtrl_bothCohorts"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$microbiome_data_type, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
genotype_differences_in_ctrls.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ genotype + cohort, 
    transform = "LOG",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE,
    plot_associations = FALSE,
    save_plots_rds = TRUE,
		verbosity = "ERROR"
  )

## add taxonomy to output tables: all_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  #arrange again based on increasing qval_individual
  arrange(qval_individual) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) 

## add taxonomy to output tables: significant_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) 

# save raw output
saveRDS(genotype_differences_in_ctrls.maaslin3, file = file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
```

## Diff. Prevalence - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
genotype_differences_in_ctrls.maaslin3_curated <- read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% 
  split.data.frame(.$model)

top20_tmp_out <- genotype_differences_in_ctrls.maaslin3_curated$prevalence %>% 
  filter(value == "Thy1-ASO") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

## Barplots top 20 most diff. prevalent

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(paste(cohort, genotype)), scales = "free_x") 
```

## Diff. Abundance - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}

top20_tmp_out <- genotype_differences_in_ctrls.maaslin3_curated$abundance %>% 
  filter(value == "Thy1-ASO") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

## Barplot top 20 most diff. abundant

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(paste(cohort, genotype)), scales = "free_x")
```

## Conclusions

There are no significantly affected microbes in Thy1-ASO mice compared to WT.

What puzzles me is that there are many taxa that are uniquely identified by each
project leader (Matheus and Ana). Also, Matheus' sample size is 10 vs 10, Ana's 
is 8 vs 8 per cohort, in total 16 vs 16. We should not be less powerful than his.

What could be the factor here?

  * facility
  
  * handlers
  
  * gavage not sterile or affecting somehow?

  * diet
  
  * Could 2 weeks make such a striking difference in the composition?

# Cohort Fp7

```{r}
cohort_subdir <- "cohort_Fp7"

dir.create(file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir), showWarnings = FALSE, recursive = TRUE)

MetaPhlAn_Fp7.tse <- MetaPhlAn.tse[, MetaPhlAn.tse@colData$cohort == "Fp7"] 
MetaPhlAn_Fp7.tse@colData$genotype_treatment <- droplevels(MetaPhlAn_Fp7.tse@colData$genotype_treatment)
```

## Overall community trends (Beta diversity)

```{r}
dimRedName <- paste("Fp7", params$beta_dist, "mds", sep = "_")
# this will work once mia or TreeSummarizedExperiment fixes some things
MetaPhlAn_Fp7.tse <- mia::addMDS(
  MetaPhlAn_Fp7.tse,
  assay.type = "metaphlan_RelAbund",
  method = params$beta_dist,
  ncomponents = 3, name = dimRedName
  )

variances_explained <- attr(reducedDim(MetaPhlAn_Fp7.tse, dimRedName), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

Fp7_colData_for_PCoA.df <- cbind.data.frame(as.data.frame(colData(MetaPhlAn_Fp7.tse)), reducedDim(MetaPhlAn_Fp7.tse, dimRedName) %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

beta_plot_basic <- Fp7_colData_for_PCoA.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = genotype_treatment_legend, fill = genotype_treatment_legend, label = cage)) + 
  geom_text() +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Genotype and Treatment",
       fill = "Genotype and Treatment",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )
```

```{r, fig.height=6, fig.width=6}
ggsave(plot = beta_plot_basic, filename = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste(params$beta_dist, params$beta_MDS, "plot.svg", sep = "_")), dpi = 600, height = 6, width = 6)

beta_plot_basic
```

Same plot as above, without WT 

```{r}
dimRedName <- paste("Fp7", params$beta_dist, "mds", sep = "_")
# this will work once mia or TreeSummarizedExperiment fixes some things
MetaPhlAn_Fp7_noWT.tse <- MetaPhlAn_Fp7.tse[,MetaPhlAn_Fp7.tse@colData$genotype != "WT"]

MetaPhlAn_Fp7_noWT.tse <- mia::addMDS(
  MetaPhlAn_Fp7_noWT.tse,
  assay.type = "metaphlan_RelAbund",
  method = params$beta_dist,
  ncomponents = 3, name = dimRedName
  )

variances_explained <- attr(reducedDim(MetaPhlAn_Fp7_noWT.tse, dimRedName), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

Fp7_colData_for_PCoA.df <- cbind.data.frame(as.data.frame(colData(MetaPhlAn_Fp7_noWT.tse)), reducedDim(MetaPhlAn_Fp7_noWT.tse, dimRedName) %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

beta_plot_basic <- Fp7_colData_for_PCoA.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = genotype_treatment_legend)) + 
  geom_point(size = 2) + 
  stat_ellipse(show.legend = FALSE) + 
  scale_color_manual(values = joe_palette) +
  theme(legend.position = "top") +
  labs(color = "Genotype and Treatment",
       x = paste0("MDS 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("MDS 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

```

```{r}
ggsave(plot = beta_plot_basic, filename = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste(params$beta_dist, params$beta_MDS, "noWT", "plot.svg", sep = "_")), dpi = 600, height = 6, width = 6)

beta_plot_basic
```

```{r}
set.seed(1234)
MetaPhlAn_Fp7.tse <- addPERMANOVA(
  MetaPhlAn_Fp7.tse,
  assay.type = "metaphlan_RelAbund",
  method = "bray",
  formula = x ~ genotype_treatment,
  name = "permanova",
  permutations = 9999,
  test.homogeneity = TRUE
)

permanova_table.df <- PERMANOVA_to_table(MetaPhlAn_Fp7.tse)

write_tsv(permanova_table.df, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste(params$beta_dist, "PERMANOVA.tsv", sep = "_")))
```

```{r}
permanova_table.df %>% 
  kbl() %>% 
  kable_styling()
```

## Pairwise group comparison (PERMANOVA and boxplot of Bray-Curtis distances)

```{r}
dissimName <- paste("Fp7", params$beta_dist, sep = "_")
MetaPhlAn_Fp7.tse <- addDissimilarity(MetaPhlAn_Fp7.tse, method = "bray", name = dissimName, assay.type= "metaphlan_RelAbund")
mtx <- metadata(MetaPhlAn_Fp7.tse)[[dissimName]]
mtx[upper.tri(mtx,diag = TRUE)] <- NA

pairwise_dissim_bray <- reshape2::melt(mtx) %>% 
  filter(complete.cases(.)) %>%
  dplyr::inner_join(., as.data.frame(colData(MetaPhlAn_Fp7.tse)) %>% transmute(Var1 = uuid, genotype_treatment_legend = genotype_treatment_legend), by = "Var1") %>% 
  dplyr::inner_join(., as.data.frame(colData(MetaPhlAn_Fp7.tse)) %>% transmute(Var2 = uuid, genotype_treatment_legend = genotype_treatment_legend), by = "Var2") %>% 
  dplyr::select(Var1, Var2, contains(c("uuid", "genotype", "treatment")), value) %>% 
  mutate(pairs = paste(genotype_treatment_legend.y, genotype_treatment_legend.x, sep = " vs ")) %>% 
  filter(genotype_treatment_legend.x != genotype_treatment_legend.y)

```

## Pairwise PERMANOVA

```{r}
source("https://raw.githubusercontent.com/pmartinezarbizu/pairwiseAdonis/refs/heads/master/pairwiseAdonis/R/pairwise.adonis.R")
set.seed(1234)
library(vegan)
pairwise_adonis_stats <- pairwise.adonis(x = as.dist(metadata(MetaPhlAn_Fp7.tse)[[dissimName]]), 
                factors = MetaPhlAn_Fp7.tse@colData$genotype_treatment, 
                perm = 9999,
                p.adjust.m = "BH"
                )

kbl(pairwise_adonis_stats, caption = "Pairwise PERMANOVA test comparing groups as seen above") %>% 
  kable_styling()

write_tsv(pairwise_adonis_stats, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste(params$beta_dist, "PERMANOVA_pairwise.tsv", sep = "_")))
```

## Same PERMANOVA statistics, but on Aitchison distance for robustness check

Input data: `MetaPhlAn counts (rel.ab x number_reads)`
Transformation: `rclr` (CLR on non-0 counts, 0 remain untouched)
Distance metric: `euclidean`

```{r}
MetaPhlAn_Fp7.tse <- transformAssay(MetaPhlAn_Fp7.tse, assay.type = "metaphlan_Counts", method = "rclr", name = "metaphlan_Counts_rclr")
set.seed(1234)
MetaPhlAn_Fp7.tse <- addPERMANOVA(
  MetaPhlAn_Fp7.tse,
  assay.type = "metaphlan_Counts_rclr",
  method = "euclidean",
  formula = x ~ genotype_treatment,
  name = "permanova_aitchison",
  permutations = 9999,
  test.homogeneity = TRUE
)
```


```{r}
# group-wise PERMANOVA
permanova_table_aitchison.df <- PERMANOVA_to_table(MetaPhlAn_Fp7.tse, name = "permanova_aitchison")

write_tsv(permanova_table_aitchison.df, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste("aitchison", "PERMANOVA.tsv", sep = "_")))
```


```{r}
# pairwise PERMANOVA
MetaPhlAn_Fp7.tse <- addDissimilarity(MetaPhlAn_Fp7.tse, method = "euclidean", name = "aitchison", assay.type= "metaphlan_Counts_rclr")

pairwise_adonis_stats_aitchison <- pairwise.adonis(x = as.dist(metadata(MetaPhlAn_Fp7.tse)[["aitchison"]]), 
                factors = MetaPhlAn_Fp7.tse@colData$genotype_treatment, 
                perm = 9999, 
                p.adjust.m = "BH"
)

write_tsv(pairwise_adonis_stats_aitchison, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, paste("aitchison", "PERMANOVA_pairwise.tsv", sep = "_")))
```

## What microbes differ between `genotypes` in Control controls?

These bacteria should not be considered in the main exposure of interest later, 
as they vary because of genotype as well. Analyses should be performed
within each cohort.

Model formula is:

$$
\log_2(Microbiome) \sim genotype
$$

```{r}
# prepare input data

tmp.tse <- MetaPhlAn_Fp7.tse[,MetaPhlAn_Fp7.tse@colData$treatment == "Control"] 
tmp.tse@colData$genotype <- relevel(tmp.tse@colData$genotype, "WT")
tmp.tse@colData$genotype <- droplevels(tmp.tse@colData$genotype)

tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "genotype_vs_WT_in_treatment_Control"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
genotype_differences_in_ctrls.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ genotype, 
    transform = "LOG",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE,
    plot_associations = TRUE,
    save_plots_rds = TRUE,
		verbosity = "ERROR"
  )

## add taxonomy to output tables: all_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  #arrange again based on increasing qval_individual
  arrange(qval_individual) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) 

## add taxonomy to output tables: significant_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) 

# save raw output
saveRDS(genotype_differences_in_ctrls.maaslin3, file = file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
```

### Diff. Prevalence - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
genotype_differences_in_ctrls.maaslin3_curated <- read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% 
  split.data.frame(.$model)

top20_tmp_out <- genotype_differences_in_ctrls.maaslin3_curated$prevalence %>% 
  filter(value == "Thy1-ASO") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplots top 20 most diff. prevalent

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(paste(treatment, genotype)), scales = "free_x") 
```

### Diff. Abundance - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
top20_tmp_out <- genotype_differences_in_ctrls.maaslin3_curated$abundance %>% 
  filter(value == "Thy1-ASO") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplot top 20 most diff. abundant

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(paste(treatment, genotype)), scales = "free_x")
```

### Conclusions

  * There are no differentially prevalent bacteria comparing Thy1-ASO vs WT 
  Control mice
  
  * 1 bacterium are decreased and 10 are increased in Thy1-ASO mice.
  
  * The large effect sizes, however, point to a lack of statistical power to 
  confidently say that there are differences. This does not mean that there are 
  0 compositional shifts
  
  * Compared to Matheus' project 1 experiment, we find much fewer associations.
  Matheus' comparison was 10 vs 10, this comparison has 8 vs 8

## Effects of `Treatment` in Thy1-ASO male mice

Model formula is:

$$
\log_2(Microbiome) \sim treament  
$$

The analysis is restricted to Thy1-ASO mice, and the baseline is Thy1-ASO-Control mice.
This is allowed because we are interested in the impact of the communities on 
mice microbiome and their motor performance, and WT mice was not gavaged either
probiotic formulation, so they are incomparable.

```{r}
# prepare input data
tmp.tse <- MetaPhlAn_Fp7.tse[,MetaPhlAn_Fp7.tse@colData$genotype != "WT"]
tmp.tse@colData$treatment <- relevel(tmp.tse@colData$treatment, "Control")
tmp.tse@colData$treatment <- droplevels(tmp.tse@colData$treatment)
tmp.tse <- tmp.tse[rowSums(assay(tmp.tse, "metaphlan_RelAbund")) > 0,]

tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "treatments_vs_Control_in_Thy1-ASO"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, maaslin3_expmt)

# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
treatments_vs_Control_in_Thy1ASO.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ treatment, 
    transform = "LOG",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE, 
    plot_associations = TRUE,
    save_plots_rds = TRUE,
		verbosity = "ERROR"
  )

## add taxonomy to output tables: all_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  #arrange again based on increasing qval_individual
  arrange(qval_individual) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) 

## add taxonomy to output tables: significant_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) 

# save raw output
saveRDS(treatments_vs_Control_in_Thy1ASO.maaslin3, file = file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
```

####  Maaslin3 default plot

```{r}
readRDS(file.path(maaslin3_final_outdir, "summary_plot_gg.RDS")) 
```

### Diff. Prevalence - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
treatments_vs_Control_in_Thy1ASO.maaslin3_curated <- read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% 
  split.data.frame(.$model)

top20_tmp_out <- treatments_vs_Control_in_Thy1ASO.maaslin3_curated$prevalence %>% 
  filter(value == "F. prausnitzii") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplots top 20 most diff. prevalent

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(genotype_treatment), scales = "free_x") 
```

### Diff. Abundance - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
top20_tmp_out <- treatments_vs_Control_in_Thy1ASO.maaslin3_curated$abundance %>% 
  filter(value == "F. prausnitzii") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplot top 20 most diff. abundant

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(paste(treatment, genotype)), scales = "free_x")
```

### Conclusions

  * No differentially **prevalent** features with Q-value < 0.1 comparing `Fp`
  against `Control` Thy1-ASO controls
  
  * No differentially **abundant** features with Q-value < 0.1 comparing `Fp`
  against `Control` Thy1-ASO controls

  * Looking at raw data, I tend to conclude the same thing: a signal is there 
  but we are underpowered to label it "statistically significant".

## Microbiome-typical plots panel

```{r}
panelA_stackplot <- plotAbundance(
  x = MetaPhlAn_Fp7.tse,
  assay.type = "metaphlan_RelAbund", 
  group = "phylum") + 
  facet_grid(cols = vars(genotype_treatment_legend), scales = "free_x") + 
  theme_light() +
  scale_x_discrete(labels = "x") +
  theme(
    panel.spacing.x = unit(0, "lines"),
    strip.background.x = element_rect(fill = "white", colour = "gray40", linewidth = unit(0.8, "mm")), 
    panel.grid = element_blank(),
    panel.border = element_rect(colour = "gray40", linewidth = unit(0.8, units = "mm")),
    strip.text = element_text(color = "black"),
    axis.text.x = element_text(color = "white"),
    axis.ticks.x = element_blank(),
    strip.text.x = element_markdown()
    ) + 
  labs(
    fill = "Phylum",
    x = " "
  )

panelA_stackplot$data$colour_by <- gsub("p__", "", panelA_stackplot$data$colour_by)
panelA_stackplot$data$colour_by <- gsub("_", " ", panelA_stackplot$data$colour_by)
```

```{r}
set.seed(1234)
dimRedName <- paste("Fp7", params$beta_dist, "mds", sep = "_")
# this will work once mia or TreeSummarizedExperiment fixes some things

MetaPhlAn_Fp7.tse <- mia::addMDS(
  MetaPhlAn_Fp7.tse,
  assay.type = "metaphlan_RelAbund",
  method = params$beta_dist,
  ncomponents = 3, name = dimRedName
  )

MetaPhlAn_Fp7.tse <- addPERMANOVA(MetaPhlAn_Fp7.tse, 
             formula = x ~ genotype_treatment, 
             assay.type = "metaphlan_RelAbund",
             method = params$beta_dist,
             permutations = 9999, 
             by = "margin")

variances_explained <- attr(reducedDim(MetaPhlAn_Fp7.tse, dimRedName), "eig")
prop_variances_explained <- variances_explained/sum(variances_explained)

Fp7_colData_for_PCoA.df <- cbind.data.frame(as.data.frame(colData(MetaPhlAn_Fp7.tse)), reducedDim(MetaPhlAn_Fp7.tse, dimRedName) %>% as.data.frame() %>% set_names(paste("MDS", 1:ncol(.))))

panelB_PCoA <- Fp7_colData_for_PCoA.df %>% 
ggplot(aes(x = `MDS 1`, y = `MDS 2`, color = genotype_treatment_legend, fill = genotype_treatment_legend)) + 
  geom_point(size = 2.5, stroke = 1, color = "black") +
  geom_point(size = 2.5) +
  ggConvexHull::geom_convexhull(show.legend = FALSE, alpha = 0.2) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  theme(legend.position = "top", legend.text = element_markdown()) +
  labs(color = "Genotype and Treatment",
       fill = "Genotype and Treatment",
       x = paste0("PCoA Axis 1 [", round(prop_variances_explained[1]*100, 1), "%]"),
       y = paste0("PCoA Axis 2 [", round(prop_variances_explained[2]*100, 2), "%]")
  )

# panelB_PCoA
```

```{r}
# function to colorize labels of ggplot object
# This function will apply ALL color rules to EACH label string
colorize_labels <- function(labels, mapping) {
  # Start with the original labels
  colored_labels <- labels
  
  # Loop through each pattern-color rule in your map
  for (pattern in names(mapping)) {
    color <- mapping[[pattern]]
    
    # The text to display is the pattern without the escape slashes
    display_text <- gsub("\\\\", "", pattern) 
    
    # Apply the coloring rule. This will modify labels that have already been partially colored.
    colored_labels <- gsub(
      pattern = pattern, 
      replacement = paste0("<span style='color:", color, ";'>", display_text, "</span>"), 
      x = colored_labels,
      fixed = TRUE
    )
  }
  
  return(colored_labels)
}

pairwise_adonis_stats_asterisks <- pairwise_adonis_stats %>% 
  mutate(pairs = gsub("Thy1-ASO F. prausnitzii", "Thy1-ASO *F. prausnitzii*", pairs))

panelC_pairwise_dissim_boxplot <- pairwise_dissim_bray %>%
  ggplot(
    aes(x = pairs, y = value)
  ) +
  geom_point(position = position_jitter(width = 0.2, seed = 1234), stroke = 1.1) +
  geom_point(position = position_jitter(width = 0.2, seed = 1234), color = "gray30") +
  geom_boxplot(outliers = FALSE, alpha = 0.6, width = 0.3) +
  geom_text(data = pairwise_adonis_stats_asterisks %>%
              dplyr::left_join(
                group_by(pairwise_dissim_bray, pairs) %>%
                  reframe(value = max(value) + 0.05), by = "pairs") %>% 
              mutate(pairs = factor(pairs, levels = pairs)),
            mapping = aes(x = pairs, y = value, label = sprintf("FDR = %.3f", p.adjusted)), hjust = 0, angle = 90) +
  # Apply the colorize_labels function to your x-axis labels
  scale_x_discrete(labels = function(x) colorize_labels(x, joe_palette)) +
  # Tell ggplot to render the axis text as Markdown/HTML
  theme(
    axis.text.x = ggtext::element_markdown(angle = 90, hjust = 1, vjust = 0.5, face = "bold")
    ) +
  ylim(0.3, 0.9) +
  labs(
    x = "",
    y = "Bray-Curtis Pairwise Dissimilarity"
  )

# make a horizontal version just in case

panelC_pairwise_dissim_boxplot_horiz <- pairwise_dissim_bray %>%
  ggplot(
    aes(y = pairs, x = value)
  ) +
  geom_point(position = position_jitter(height = 0.2, seed = 1234), stroke = 1.1) +
  geom_point(position = position_jitter(height = 0.2, seed = 1234), color = "gray30") +
  geom_boxplot(outliers = FALSE, alpha = 0.6, width = 0.3) +
  geom_text(data = pairwise_adonis_stats_asterisks %>%
              dplyr::left_join(
                group_by(pairwise_dissim_bray, pairs) %>%
                  reframe(value = max(value) + 0.05), by = "pairs") %>% 
              mutate(pairs = factor(pairs, levels = pairs)),
            mapping = aes(y = pairs, x = value, label = sprintf("FDR = %.3f", p.adjusted)), hjust = 0) +
  # Apply the colorize_labels function to your x-axis labels
  scale_y_discrete(labels = function(x) colorize_labels(x, joe_palette)) +
  # Tell ggplot to render the axis text as Markdown/HTML
  theme(axis.text.y = ggtext::element_markdown(angle = 0, hjust = 1)) +
  xlim(0.3, 0.9) +
  labs(
    y = "",
    x = "Bray-Curtis Pairwise Dissimilarity"
  )

```

```{r}
alpha_data.df <- MetaPhlAn_Fp7.tse %>%
  colData() %>%
  as.data.frame() %>% 
  dplyr::rename(
    `Shannon Diversity` = shannon_diversity,
    `Observed Richness` = observed_richness
  )
comparisons <- combn(levels(alpha_data.df$genotype_treatment_legend), 2, simplify = FALSE)

stats_obs <- rstatix::pairwise_wilcox_test(alpha_data.df, formula = `Observed Richness` ~ genotype_treatment,p.adjust.method =  "BH") %>% 
  mutate(group2 = gsub("F. prausnitzii", "*F. prausnitzii*", group2),
         p.adj.round3 = round(p.adj, 3)) %>% 
  rstatix::add_y_position(step.increase = 0.5)

# Observed richness diversity vs genotype-treatment
panelD_Obs_boxplot <- alpha_data.df %>% 
  ggplot(aes(x = genotype_treatment_legend, y = `Observed Richness`, color = genotype_treatment_legend, fill = genotype_treatment_legend)) + 
  geom_boxplot(outliers = FALSE, show.legend = FALSE, color = "black", alpha = 0.4) + 
  geom_point(color = "black", stroke = 1.1, position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  geom_point(position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  stat_pvalue_manual(stats_obs, label = "p.adj.round3", tip.length = 0.01) + 
  theme(axis.text.x = element_blank(), legend.text = element_markdown()) +
  labs(
    x = NULL,
    color = "Genotype and Treatment",
    fill = "Genotype and Treatment"
  )

stats_shannon <- rstatix::pairwise_wilcox_test(alpha_data.df, formula = `Shannon Diversity` ~ genotype_treatment, p.adjust.method =  "BH") %>% 
  mutate(group2 = gsub("F. prausnitzii", "*F. prausnitzii*", group2),
         p.adj.round3 = round(p.adj, 3)) %>% 
  rstatix::add_y_position(step.increase = 0.5)

# Shannon diversity vs genotype-treatment
panelE_Shannon_boxplot <- alpha_data.df %>% 
  ggplot(aes(x = genotype_treatment_legend, y = `Shannon Diversity`, color = genotype_treatment_legend, fill = genotype_treatment_legend)) + 
  geom_boxplot(outliers = FALSE, show.legend = FALSE, color = "black", alpha = 0.4) + 
  geom_point(color = "black", stroke = 1.1, position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  geom_point(position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  stat_pvalue_manual(stats_shannon, label = "p.adj.round3", tip.length = 0.01) + 
  theme(axis.text.x = element_blank(), legend.text = element_markdown()) +
  labs(
    x = NULL,
    color = "Genotype and Treatment",
    fill = "Genotype and Treatment"
  )

# alpha_plots_panel <- ggarrange(panelE_Obs.gg, panelE_Shannon_boxplot)

```

```{r}

panelF_FB.Ratio.df <- MetaPhlAn_Fp7.tse %>%
  colData() %>%
  as.data.frame()

stats_fbrartio <- rstatix::pairwise_wilcox_test(panelF_FB.Ratio.df, formula = FB.Ratio ~ genotype_treatment, p.adjust.method =  "BH") %>% 
  mutate(group2 = gsub("F. prausnitzii", "*F. prausnitzii*", group2),
         p.adj.round3 = round(p.adj, 3)) %>% 
  rstatix::add_y_position(step.increase = 0.5)


panelF_FB.Ratio <- panelF_FB.Ratio.df %>% 
  ggplot(aes(x = genotype_treatment_legend, y = FB.Ratio, color = genotype_treatment_legend, fill = genotype_treatment_legend)) + 
  geom_boxplot(outliers = FALSE, show.legend = FALSE, color = "black", alpha = 0.4) + 
  geom_point(color = "black", stroke = 1.1, position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  geom_point(position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  stat_pvalue_manual(stats_fbrartio, label = "p.adj.round3", tip.length = 0.01) + 
  theme(axis.text.x = element_blank(), legend.text = element_markdown()) +
  labs(
    x = NULL,
    y = "Firmicutes/Bacteroidetes Ratio",
    color = "Genotype and Treatment",
    fill = "Genotype and Treatment"
  )

max_fb_ratio <- max(panelF_FB.Ratio.df$FB.Ratio, na.rm = TRUE)

panelF_FB.Ratio_horiz <- panelF_FB.Ratio.df %>% 
  mutate(genotype_treatment_legend = factor(genotype_treatment_legend, levels = rev(levels(genotype_treatment_legend)))) %>% 
  ggplot(aes(x = genotype_treatment_legend, y = FB.Ratio, color = genotype_treatment_legend, fill = genotype_treatment_legend)) + 
  geom_boxplot(outliers = FALSE, show.legend = FALSE, color = "black", alpha = 0.4) + 
  geom_point(color = "black", stroke = 1.1, position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  geom_point(position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  coord_flip() + 
  stat_pvalue_manual(stats_fbrartio, label = "p.adj.round3", tip.length = 0.01, coord.flip = TRUE) + 
  theme(axis.text.y = element_blank(), legend.text = element_markdown()) +
  labs(
    x = NULL,
    y = "Firmicutes/Bacteroidetes Ratio",
    color = "Genotype and Treatment",
    fill = "Genotype and Treatment"
  )
```

```{r}
# panel C, differentially abundant features, or top 20 more strongly associated

GTDB.df <- GTDB.df %>%
  mutate(
    altNames = sapply(
      strsplit(paste(trimws(Genus), trimws(Species), sprintf("[%s]", SGB)), " ", fixed = TRUE),
      function(x) paste(unique(x), collapse = " ")
    )
  )

altNames.chr <- GTDB.df$altNames
names(altNames.chr) <- GTDB.df$SGB

panelG_diffAbund <- readRDS("results/SGB/cohort_Fp7/treatments_vs_Control_in_Thy1-ASO/summary_plot_gg.RDS") + 
  theme_light() + 
  theme(strip.text = element_blank()) + 
  # expand plot to show full errorbars
  scale_x_continuous(expand = expansion(mult = c(.1))) + 
  # remove the null HP that would mess with the panel rendering
  guides(linetype = "none")


panelG_diffAbund2 <- panelG_diffAbund +
  scale_y_discrete(
    labels = function(y) altNames.chr[gsub("t__", "", y)]
  )

```

```{r, fig.height=12, fig.width=16}
#' panelA_stackplot
#' panelB_PCoA
#' panelC_pairwise_dissim_boxplot
#' pairwise_dissim_boxplot
#' panelD_Obs_boxplot
#' panelE_Shannon_boxplot
#' panelF_FB.Ratio
#' panelG_diffAbund

piece1 <- panelA_stackplot + theme(legend.position = "top") + guides(fill = guide_legend(ncol = 2, title = NULL))

piece2 <- ggarrange(panelB_PCoA, panelC_pairwise_dissim_boxplot, common.legend = TRUE, widths = c(1, 0.5), labels = c("B", "C"))

piece3 <- ggarrange(
  panelD_Obs_boxplot + theme(legend.position = "none"), 
  panelE_Shannon_boxplot + theme(legend.position = "none"),
  panelF_FB.Ratio + theme(legend.position = "none"), nrow = 1, labels = c("D", "E", "F")
  )

piece2_3 <- ggarrange(piece2, piece3, nrow = 2)

piece_1_2_3 <- ggarrange(piece1, piece2_3, nrow = 1, ncol = 2, widths = c(0.6, 1),  labels = c("A", ""))

piece_1_2_3_4 <- ggarrange(piece_1_2_3, panelG_diffAbund, widths = c(1, 0.6), labels = c("", "G"))

ggsave(plot = piece_1_2_3_4, filename = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, "microbiome_panel_Full.svg"), width = 18, height = 10)

piece_1_2_3_4
```

```{r,fig.height=8, fig.width=12}
panel_A_legend <- get_legend(panelA_stackplot + guides(fill = guide_legend(nrow = 2, title = "Phylum", position = "top")))

microbiome_panel_MainText <- ggarrange(

  ggarrange(
    panelA_stackplot + guides(fill = guide_legend(title = "Phylum")) + labs(y = "Relative Abundance", x = "Sample") + theme(axis.title = element_text(size = 12)), 
    panelB_PCoA + theme(axis.title = element_text(size = 12)), 
    common.legend = TRUE, widths = c(1, 1), legend = "top", labels = c("A", "B")),
    
    ggarrange(
      panelC_pairwise_dissim_boxplot_horiz  + theme(axis.title = element_text(size = 12), axis.text = element_text(size = 9)),
      get_legend(panelB_PCoA + guides(color = guide_legend(ncol = 1, position = "right"))), 
      
      nrow = 1, widths = c(1,0.2), labels = c("C", "")
      ),
  
  panelG_diffAbund2 + 
    coord_flip() + 
    theme(axis.text.x = element_text(angle = 90, hjust = 1, vjust = 1), 
          legend.position = "bottom") + 
    labs(x = expression(log[2]~Fold~Change), y = NULL) + 
    theme(axis.title = element_text(size = 12)),
  
  nrow = 3, 
  heights = c(1, 0.25, 1), labels = c("", "", "D"))

ggsave(plot = microbiome_panel_MainText, filename = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, "microbiome_panel_MainText.svg"), width = 11, height = 13)

microbiome_panel_MainText

# get_legend(panelB_PCoA)
```

```{r}
microbiome_panel_Supplementary <- ggarrange(
    panelD_Obs_boxplot + theme(axis.ticks.x = element_blank()) + scale_y_continuous(expand = expansion(mult = 0.08)),
    panelE_Shannon_boxplot + theme(axis.ticks.x = element_blank()) + scale_y_continuous(expand = expansion(mult = 0.08)),
    panelF_FB.Ratio + theme(axis.ticks.x = element_blank()) + scale_y_continuous(expand = expansion(mult = 0.08)),
    nrow = 1, labels = c("A", "B", "C"), 
common.legend = TRUE)

ggsave(plot = microbiome_panel_Supplementary, filename = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, "microbiome_panel_Supplementary.svg"), width = 10, height = 4)

microbiome_panel_Supplementary
```

## Treatment effect on Motor and GI symptoms (selection of variables)

```{r}
# gather bead_expulsion_n data
bead_expulsion.df <- openxlsx::read.xlsx("Data/motor_performance/Fp7_raw.xlsx", sheet = "Bead_exp") %>% 
  dplyr::transmute(
    mouse_id = ID,
    bead_expulsion_min = Trial1)

# gather beam_over_steps ratio for trial 1 specifically, as presented in the rest of the publication
slips_over_step.df <- openxlsx::read.xlsx("Data/motor_performance/Fp7_raw.xlsx", sheet = "Beam_steps") %>% 
  rowwise() %>% 
  transmute(
    mouse_id = ID,
    Slips = Slips_Trial1,
    Steps = Steps_Trial1,
    # Slips = mean(c(Slips_Trial1, Slips_Trial2), na.rm = TRUE),
    # Steps = mean(c(Steps_Trial1, Steps_Trial2), na.rm = TRUE),
    slips_over_steps = Slips/Steps)

additional_variables.df <- dplyr::full_join(bead_expulsion.df, slips_over_step.df, by = "mouse_id")

# incorporate these columns into the colData of the input
new_df <- DataFrame(dplyr::left_join(as.data.frame(colData(MetaPhlAn_Fp7.tse)), additional_variables.df, by = "mouse_id"))
rownames(new_df) <- new_df$uuid

colData(MetaPhlAn_Fp7.tse) <- new_df
```

```{r, fig.height=4, fig.width=7}
motor_variables <- grepv("bead_expulsion_min|slips_over_steps", colnames(colData(MetaPhlAn_Fp7.tse)))

comparisons <- combn(levels(colData(MetaPhlAn_Fp7.tse)$genotype_treatment_legend), 2, simplify = FALSE)

input.df <- as.data.frame(colData(MetaPhlAn_Fp7.tse))

motor_performance_per_group.gg.list <- sapply(motor_variables, function(mot) {
  na.logical <- is.na(input.df[[mot]])
  ggplot(
    input.df %>% filter(!na.logical),
         aes(x = genotype_treatment, y = eval(parse(text = mot)), color = genotype_treatment, fill = genotype_treatment)) + 
  geom_boxplot(outliers = FALSE, show.legend = FALSE, color = "black", alpha = 0.4) + 
  geom_point(color = "black", stroke = 1.1, position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  geom_point(position= position_jitterdodge(jitter.width = 0.5, seed = 1234)) + 
  scale_color_manual(values = joe_palette) +
  scale_fill_manual(values = joe_palette) +
  stat_compare_means(
    comparisons = comparisons, method = "wilcox.test", method.args = list(exact = FALSE)
  )+ 
  theme(legend.position = "none", axis.text.x = element_blank()) +
  labs(
    x = "",
    y = mot,
    caption = sprintf("N. NAs: %d", sum(na.logical))
  )
}, USE.NAMES = TRUE, simplify = FALSE)

ggarrange(plotlist = motor_performance_per_group.gg.list, common.legend = TRUE)

```

### Mediation analysis on `slips_over_steps` {.tabset}

```{r}
library(mediation)

# simple function to tidy a mediation analysis object
tidy_mediation <- function(mediation.model){
  med_summary <- summary(mediation.model)
  results_df <- data.frame(
    Effect = c("ACME", "ADE", "Total Effect", "Prop. Mediated"),
    Estimate = c(med_summary$z.avg, med_summary$d.avg, med_summary$tau.coef, med_summary$n.avg),
    CI_Lower_95 = c(med_summary$z.avg.ci[1], med_summary$d.avg.ci[1], med_summary$tau.ci[1], med_summary$n.avg.ci[1]),
    CI_Upper_95 = c(med_summary$z.avg.ci[2], med_summary$d.avg.ci[2], med_summary$tau.ci[2], med_summary$n.avg.ci[2]),
    P_Value = c(med_summary$z.avg.p, med_summary$d.avg.p, med_summary$tau.p, med_summary$n.avg.p)
  )
  
  return(results_df)
}


mediationData.df <- cbind.data.frame(
  as.data.frame(colData(MetaPhlAn_Fp7.tse)), 
as.data.frame(reducedDim(MetaPhlAn_Fp7.tse, dimRedName))) %>% 
  dplyr::rename(
    PCo.1 = V1, 
    PCo.2 = V2,
    PCo.3 = V3
    ) %>% 
  filter(
    # remove NAs from covariates of interest
    complete.cases(dplyr::select(., contains("PCo"), genotype_treatment, slips_over_steps)), 
    # remove WT Control and use ASO control as baseline
    genotype_treatment != "WT Control"
    ) %>% 
  mutate(
    treatment = droplevels(treatment)
  )

# make 3 mediation analyses as list, one per Principal Component
mediations_micrombiome <- sapply(grepv("PCo", colnames(mediationData.df)), function(x) {
  mediation.lm <- lm(as.formula(paste(x, "treatment", sep = "~")), data = mediationData.df)
  outcome.lm <- lm(as.formula(paste("slips_over_steps ~ treatment",  x, sep = "+")), data = mediationData.df)
  return(
    list(
      mediator.lm = mediation.lm,
      outcome.lm = outcome.lm,
      mediation.analysis = mediate(
        model.m = mediation.lm,
        model.y = outcome.lm,
        sims = 1000,
        treat = "treatment",
        mediator = x,
        boot = FALSE,
        robustSE = TRUE
      )
    )
  )
}, simplify = FALSE, USE.NAMES=TRUE)

data_as_long_table <- mediations_micrombiome %>% 
  lapply("[[", "mediation.analysis") %>% 
  lapply(tidy_mediation) %>% 
  bind_rows(.id = "PCo")

write_tsv(data_as_long_table, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, "PCo1-3_mediation_slips_over_steps_ASO_treat_vs_ASO_ctrl.tsv"))
```

#### PCo 1

```{r}
summary(mediations_micrombiome$PCo.1$mediation.analysis)
```

#### PCo 2

```{r}
summary(mediations_micrombiome$PCo.2$mediation.analysis)
```

#### PCo 3

```{r}
summary(mediations_micrombiome$PCo.3$mediation.analysis)
```

### {-}

### Mediation analysis on `bead_expulsion_min` {.tabset}

```{r}
library(mediation)

# simple function to tidy a mediation analysis object
tidy_mediation <- function(mediation.model){
  med_summary <- summary(mediation.model)
  results_df <- data.frame(
    Effect = c("ACME", "ADE", "Total Effect", "Prop. Mediated"),
    Estimate = c(med_summary$z.avg, med_summary$d.avg, med_summary$tau.coef, med_summary$n.avg),
    CI_Lower_95 = c(med_summary$z.avg.ci[1], med_summary$d.avg.ci[1], med_summary$tau.ci[1], med_summary$n.avg.ci[1]),
    CI_Upper_95 = c(med_summary$z.avg.ci[2], med_summary$d.avg.ci[2], med_summary$tau.ci[2], med_summary$n.avg.ci[2]),
    P_Value = c(med_summary$z.avg.p, med_summary$d.avg.p, med_summary$tau.p, med_summary$n.avg.p)
  )
  
  return(results_df)
}


mediationData.df <- cbind.data.frame(
  as.data.frame(colData(MetaPhlAn_Fp7.tse)), 
as.data.frame(reducedDim(MetaPhlAn_Fp7.tse, dimRedName))) %>% 
  dplyr::rename(
    PCo.1 = V1, 
    PCo.2 = V2,
    PCo.3 = V3
    ) %>% 
  filter(
    # remove NAs from covariates of interest
    complete.cases(dplyr::select(., contains("PCo"), genotype_treatment, slips_over_steps)), 
    # remove WT Control and use ASO control as baseline
    genotype_treatment != "WT Control"
    ) %>% 
  mutate(
    treatment = droplevels(treatment)
  )

mediations_micrombiome <- sapply(grepv("PCo", colnames(mediationData.df)), function(x) {
  mediation.lm <- lm(as.formula(paste(x, "treatment", sep = "~")), data = mediationData.df)
  outcome.lm <- lm(as.formula(paste("slips_over_steps ~ treatment",  x, sep = "+")), data = mediationData.df)
  return(
    list(
      mediator.lm = mediation.lm,
      outcome.lm = outcome.lm,
      mediation.analysis = mediate(
        model.m = mediation.lm,
        model.y = outcome.lm,
        sims = 1000,
        treat = "treatment",
        mediator = x,
        boot = FALSE,
        robustSE = TRUE
      )
    )
  )
}, simplify = FALSE, USE.NAMES=TRUE)

data_as_long_table <- mediations_micrombiome %>% 
  lapply("[[", "mediation.analysis") %>% 
  lapply(tidy_mediation) %>% 
  bind_rows(.id = "PCo")

write_tsv(data_as_long_table, file = file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, "PCo1-3_mediation_bead_expulsion_min_ASO_treat_vs_ASO_ctrl.tsv"))
```

#### PCo 1

```{r}
summary(mediations_micrombiome$PCo.1$mediation.analysis)
```

#### PCo 2

```{r}
summary(mediations_micrombiome$PCo.2$mediation.analysis)
```

#### PCo 3

```{r}
summary(mediations_micrombiome$PCo.3$mediation.analysis)
```

### {-}

# Track probiotic species across all samples

## Fp (F. prausnitzii)

```{r}
Fp_species <- "s__Faecalibacterium_prausnitzii"
```


```{r}
tmp.tse <- MetaPhlAn.tse
rownames(tmp.tse) <- collapse_taxonomy(as.data.frame(rowData(tmp.tse)))

tmp.tse %>% 
  meltSE(assay.type = "metaphlan_RelAbund", add.col = c("cohort", "genotype", "treatment")) %>%
  filter(grepl(paste(Fp_species, collapse = "|"), FeatureID)) %>% 
  group_by(FeatureID, genotype, cohort, treatment) %>% 
  reframe(
    mean_RelAbund = mean(metaphlan_RelAbund)
  ) %>% 
  DT::datatable(
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
  
```

## bCL (benCom community)

```{r}
benCom_species <- c("s__Roseburia_intestinalis", "s__Roseburia_faecis", "s__Fusicatenibacter_saccharivorans", "s__Bacteroides_vatus", "s__Eubacterium_rectale" , "s__Anaerostipes_hadrus", "s__Faecalibacterium_prausnitzii", "Prevotella_histicola")

benCom_species
```


```{r}
tmp.tse <- MetaPhlAn.tse
rownames(tmp.tse) <- collapse_taxonomy(as.data.frame(rowData(tmp.tse)))

tmp.tse %>% 
  meltSE(assay.type = "metaphlan_RelAbund", add.col = c("cohort", "genotype", "treatment")) %>%
  filter(grepl(paste(benCom_species, collapse = "|"), FeatureID)) %>% 
  group_by(FeatureID, genotype, cohort, treatment) %>% 
  reframe(
    mean_RelAbund = mean(metaphlan_RelAbund)
  ) %>% 
  DT::datatable(
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

## Conclusions

    * No species from the Fp or benCom community was found
    
    * Time between sampling and last gavage?
    
    * Diet low in fibers?
    
    * Phylogenetically close bacteria may have taken the niche or serve a 
    similar purpose (i.e. butyrate production)?

# Are there cohort-specific bacteria in the dataset?

This can be asked on all data adjusting for other effects **OR** Comparing 
WT mice between the two cases (cleaner but low statistical power). 
I decided to do the first. 

$$
\log_2(Microbiome) \sim genotype + treatment + cohort
$$

```{r}
# prepare input data

tmp.tse <- MetaPhlAn.tse 

tmp.tse

# define experiment name and output directories
maaslin3_expmt <- "cohort_assoc_bacteria"
maaslin3_final_outdir <- file.path(params$basic_outdir, params$microbiome_data_type, cohort_subdir, maaslin3_expmt)
# create definitive output directory
dir.create(maaslin3_final_outdir, showWarnings = FALSE, recursive = TRUE)
```

Results are stored in ``r maaslin3_final_outdir``

```{r}
set.seed(1234)
cohort_differences.maaslin3 <- maaslin3(
    input_data = tmp.tse,
    formula = ~ genotype + treatment + cohort, 
    transform = "LOG",
    output = maaslin3_final_outdir,
    # do not standardize continuous metadata variables in the model
    standardize = FALSE, 
    plot_associations = TRUE,
    save_plots_rds = TRUE,
		verbosity = "ERROR"
  )

## add taxonomy to output tables: all_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  #arrange again based on increasing qval_individual
  arrange(qval_individual) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) 

## add taxonomy to output tables: significant_results.tsv
dplyr::full_join(as.data.frame(rowData(tmp.tse)), read_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) %>% dplyr::rename(!!sym(params$microbiome_data_type) := feature), by = params$microbiome_data_type) %>% 
  write_tsv(file.path(maaslin3_final_outdir, "significant_results.tsv")) 

# save raw output
saveRDS(cohort_differences.maaslin3, file = file.path(maaslin3_final_outdir, "maaslin3_raw_output.rds"))
```

### Diff. Prevalence - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
cohort_differences.maaslin3_curated <- read_tsv(file.path(maaslin3_final_outdir, "all_results.tsv")) %>% 
  split.data.frame(.$model)

top20_tmp_out <- cohort_differences.maaslin3_curated$prevalence %>% 
  filter(value == "Fp7") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplots top 20 most diff. prevalent

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(cohort), scales = "free_x") 
```

### Diff. Abundance - Top 20 with lowest crude p-value and highest absolute $\beta$

```{r}
top20_tmp_out <- cohort_differences.maaslin3_curated$abundance %>% 
  filter(value == "Fp7") %>% 
  arrange(pval_individual, desc(abs(coef))) %>% 
  head(20)

DT::datatable(
    top20_tmp_out,
     extensions = "Buttons",
    options = list(dom = "Bfrtip",
                   buttons = c('copy', 'csv', 'excel', 'pdf'))
  )
```

### Barplot top 20 most diff. abundant

```{r, fig.height=6, fig.width=20}

tmpTop20.tse <- tmp.tse[top20_tmp_out$SGB,]
rownames(tmpTop20.tse) <- collapse_taxonomy(as.data.frame(rowData(tmpTop20.tse)))

miaViz::plotAbundance(tmpTop20.tse, assay.type = "metaphlan_RelAbund", group = "SGB") + 
  facet_grid(cols = vars(cohort), scales = "free_x")
```

# Enterobacteria abundances in relation to the treatment

```{r}
enteros_SGBs <- rowData(MetaPhlAn.tse) %>% 
  as.data.frame() %>% 
  filter(family == "f__Enterobacteriaceae") %>% 
  .$SGB

# set names of SGBs as specie
names(enteros_SGBs) <- rowData(MetaPhlAn.tse) %>% 
  as.data.frame() %>% 
  filter(family == "f__Enterobacteriaceae") %>% 
  .$species

enteros_SGBs_top4 <- head(enteros_SGBs, 4)

enteros_SGBs_top4
```

```{r}
tmpDF <- meltSE(MetaPhlAn.tse, assay.type = "metaphlan_RelAbund", add.col = c("cohort", "genotype", "treatment")) %>% 
  filter(FeatureID == "t__SGB10068") %>% 
    mutate(Treatment = factor(ifelse(treatment == "Control", paste(genotype, "Control", sep = "\n"), paste(genotype, "Treated", sep = "\n")), levels = c("WT\nControl", "Thy1-ASO\nControl", "Thy1-ASO\nTreated")))


e.coli_relAbund_plot <- tmpDF %>% 
  ggplot(aes(x = cohort, y = metaphlan_RelAbund, color = Treatment)) +
  geom_point(position = position_jitterdodge(dodge.width = 0.8, seed = 1234), size = 2) +
  geom_boxplot(alpha = 0.5,
               show.legend = FALSE,
               outliers = FALSE) +
  theme(legend.position = "top", axis.title.y = element_text(face = "italic")) +
  labs(x = "Treatment Cohort", color = "Genotype\nTreatment", y = "E. coli", caption = "FDR from pairwise Wilcoxon Test")
  
e.coli_relAbund_plot
```

# Overall Conclusions (to be done)

# Save Supplementary tables

## Default taxonomy table

```{r}
write_tsv(as_tibble(rowData(MetaPhlAn.tse)), file = file.path(params$basic_outdir, params$microbiome_data_type, "Supplementary table 1 - taxonomy.tsv"))
```

# GTDB taxonomy table (only relevant bacteria)

```{r}
write_tsv(filter(GTDB.df, SGB %in% gsub("t__", "", rowData(MetaPhlAn.tse)$SGB)), file = file.path(params$basic_outdir, params$microbiome_data_type, "Supplementary table 2 - taxonomy GTDB.tsv"))
```


# Session Info

```{r}
sessionInfo()
```

  